pytato/transform.py

```diff
@@ -660,4 +671,390 @@ def get_dependencies(expr: DictOfNamedArrays) -> Dict[str, FrozenSet[Array]]:
+
+    def map_concatenate(self, expr: Concatenate, *args: Any) -> Concatenate:
+        new_bindings: List[Array] = []
+        for c in expr.arrays:
+            new_bindings.append(self._handle_new_binding(expr, c))
+
+        return Concatenate(
+                     arrays=tuple(new_bindings),
+                     axis=expr.axis,
+                     tags=expr.tags)
+
+    def map_stack(self, expr: Stack, *args: Any) -> Stack:
+        new_bindings: List[Array] = []
+        for c in expr.arrays:
+            new_bindings.append(self._handle_new_binding(expr, c))
+
+        return Stack(
+                     arrays=tuple(new_bindings),
+                     axis=expr.axis,
+                     tags=expr.tags)
+
+    def map_roll(self, expr: Roll, *args: Any) -> Roll:
+        new_binding = self._handle_new_binding(expr, expr.array)
+
+        return Roll(array=new_binding,
+                shift=expr.shift,
+                axis=expr.axis,
+                tags=expr.tags)
+
+    def map_placeholder(self, expr: Placeholder, *args: Any) -> Placeholder:
+        new_shapes: List[Array] = []
+        for dim in expr.shape:
+            if isinstance(dim, Array):
+                new_shapes.append(self._handle_new_binding(expr, dim))
+
+        assert expr.name
+
+        return Placeholder(name=expr.name,
+                shape=tuple(new_shapes),
+                dtype=expr.dtype,
+                tags=expr.tags)
+
+    def map_matrix_product(self, expr: MatrixProduct, *args: Any) -> MatrixProduct:
+
+        new_x1 = self._handle_new_binding(expr, expr.x1)
+        new_x2 = self._handle_new_binding(expr, expr.x2)
+
+        return MatrixProduct(x1=new_x1, x2=new_x2,
+                tags=expr.tags)
+
+    def map_index_lambda(self, expr: IndexLambda, *args: Any) -> IndexLambda:
+        new_bindings: Dict[str, Array] = {}
+        for name, child in expr.bindings.items():
+            new_bindings[name] = self._handle_new_binding(expr, child)
+
+        new_shapes: List[Array] = []
+        for dim in expr.shape:
+            if isinstance(dim, Array):
+                new_shapes.append(self._handle_new_binding(expr, dim))
+
+        return IndexLambda(expr=expr.expr,
+                shape=tuple(new_shapes),
+                dtype=expr.dtype,
+                bindings=new_bindings,
+                tags=expr.tags)
+
+    def __call__(self, expr: Array, *args: Any, **kwargs: Any) -> Array:
+        return self.rec(expr)
+
+
+from pytato.target import BoundProgram
+
+
+@dataclass
+class CodePartitions:
+    """Store partitions and their code."""
+    toposorted_partitions: List[Hashable]
+    prg_per_partition: Dict[Hashable, BoundProgram]
+    partition_id_to_input_names: Dict[Hashable, List[str]]
+    partition_id_to_output_names: Dict[Hashable, List[str]]
+
+
+def find_partitions(expr: Array, part_func: Callable[[Array], Hashable]) ->\
+        CodePartitions:
+    """Partitions the `expr` according to `part_func` and generates code for
```

Comment:
```suggestion
    """Partitions the *expr* according to *part_func* and generates code for
```
(according to sphinx style guide)